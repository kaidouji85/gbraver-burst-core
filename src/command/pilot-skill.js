// @flow

/**
 * パイロットスキルコマンド
 */
export type PilotSkillCommand = {
  type: 'PilotSkillCommand'
}