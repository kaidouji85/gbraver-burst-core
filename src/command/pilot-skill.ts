/**
 * パイロットスキルコマンド
 */
export type PilotSkillCommand = {
  type: "PILOT_SKILL_COMMAND";
};