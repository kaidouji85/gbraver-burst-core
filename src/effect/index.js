// @flow

import type {InputCommand} from "./input-command/input-command";
import type {Battle} from "./battle/battle";
import type {TurnChange} from "./turn-change/turn-change";
import type {StartGame} from "./start-game/start-game";
import type {BurstEffect} from "./burst/burst-effect";
import type {GameEnd} from "./game-end/game-end";
import type {BatteryDeclaration} from "./battery-declaration/battery-declaration";
import type {Reflect} from "./reflect/reflect";
import type {UpdateRemainingTurn} from "./update-remaning-turn/update-remaining-turn";

/** ゲーム中に発生する効果をまとめたもの */
export type Effect =
  StartGame |
  GameEnd |
  InputCommand |
  Reflect |
  BatteryDeclaration |
  Battle |
  TurnChange |
  BurstEffect |
  UpdateRemainingTurn;
