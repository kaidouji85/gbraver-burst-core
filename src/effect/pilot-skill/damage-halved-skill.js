// @flow

import type { DamageHalvedSkill } from "../../player/pilot";
import type { PlayerId } from "../../player/player";
import type { GameState, GameStateX } from "../../state/game-state";
import type { PlayerState } from "../../state/player-state";
import type { PilotSkillEffectX } from "./pilot-skill-effect";

/**
 * パイロットスキル ダメージ半減
 * @param lastState 最新のステート
 * @param invokerId スキル発動者のID
 * @param skill スキル内容
 * @return 更新結果、実行不可能な場合は例外を投げる
 */
export function damageHalvedSkill(
  lastState: GameState,
  invokerId: PlayerId,
  skill: DamageHalvedSkill
): GameStateX<PilotSkillEffectX<DamageHalvedSkill>> {
  const invoker = lastState.players.find((v) => v.playerId === invokerId);
  if (!invoker) {
    throw new Error("not found pilot skill invoker");
  }

  const updatedInvoker = {
    ...invoker,
    armdozer: {
      ...invoker.armdozer,
      effects: [
        ...invoker.armdozer.effects,
        {
          type: "DamageHalved",
          period: { type: "TurnLimit", remainingTurn: skill.duration },
        },
      ],
    },
  };
  const updatedPlayers: PlayerState[] = lastState.players.map((v) =>
    v.playerId === invokerId ? updatedInvoker : v
  );
  const effect = {
    name: "PilotSkillEffect",
    invokerId: invokerId,
    skill,
  };
  return { ...lastState, players: updatedPlayers, effect };
}
