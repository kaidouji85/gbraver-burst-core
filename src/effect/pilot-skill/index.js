// @flow

import type {GameState, GameStateX, PilotSkill, PilotSkillEffect, PilotSkillEffectX, PlayerId} from "../..";
import type {BuffPowerSkill, RecoverBatterySkill} from "../../player/pilot";
import {recoverBattery} from "./recover-battery";
import {buffPower} from "./buff-power";

/**
 * パイロットスキルを発動する
 *
 * @param lastState 最新の状態
 * @param invokerId パイロットスキルを発動するプレイヤー
 * @return 更新結果、実行不可能な場合はnullを返す
 */
export function pilotSkill(lastState: GameState, invokerId: PlayerId): ?GameStateX<PilotSkillEffect> {
  const donePilotSkill = pilotSkillEffect(lastState, invokerId);
  if (!donePilotSkill) {
    return null;
  }
  return disablePilotSkill(donePilotSkill);
}

/**
 * パイロットスキルを適用する
 *
 * @param lastState 最新の状態
 * @param invokerId パイロットスキルを発動するプレイヤー
 * @return 更新結果、実行不可能な場合はnullを返す
 */
function pilotSkillEffect(lastState: GameState, invokerId: PlayerId): ?GameStateX<PilotSkillEffect> {
  const invoker = lastState.players.find(v => v.playerId === invokerId);
  if (!invoker) {
    return null;
  }

  if (invoker.pilot.skill.type === 'RecoverBatterySkill') {
    const castedSkill: RecoverBatterySkill = invoker.pilot.skill;
    const updated =  recoverBattery(lastState, invokerId, castedSkill);
    return updated ? upcastPilotSkillEffect(updated) : null;
  }

  if (invoker.pilot.skill.type === 'BuffPowerSkill') {
    const castedSkill: BuffPowerSkill = invoker.pilot.skill;
    const updated = buffPower(lastState, invokerId, castedSkill);
    return updated ? upcastPilotSkillEffect(updated) : null;
  }

  return null;
}

/**
 * パイロットスキルを使用済みにする
 *
 * @param lastState 最新状態
 * @return 更新結果、実行不可能な場合はnullを返す
 */
function disablePilotSkill(lastState: GameStateX<PilotSkillEffect>): ?GameStateX<PilotSkillEffect> {
  const invoker = lastState.players.find(v => v.playerId === lastState.effect.invokerId);
  if (!invoker) {
    return null;
  }

  const updatedInvoker = {
    ...invoker,
    pilot: {
      ...invoker.pilot,
      enableSkill: false
    }
  };
  const updatedPlayers = lastState.players
    .map(v => v.playerId === updatedInvoker.playerId ? updatedInvoker : v);
  return {
    ...lastState,
    players: updatedPlayers
  };
}

/**
 * パイロットスキル発動ステートにアップキャストする
 *
 * @param origin キャスト前
 * @return キャスト結果
 */
function upcastPilotSkillEffect<X: PilotSkill>(origin: GameStateX<PilotSkillEffectX<X>>): GameStateX<PilotSkillEffect> {
  return ((origin: any): GameStateX<PilotSkillEffectX<PilotSkill | typeof origin.effect.skill>>);
}
