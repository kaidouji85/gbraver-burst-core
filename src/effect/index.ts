import { z } from "zod";

import { BatteryDeclaration, BatteryDeclarationSchema } from "./battery-declaration/battery-declaration";
import { Battle, BattleSchema } from "./battle/battle";
import { BurstEffect, BurstEffectSchema } from "./burst/burst-effect";
import { GameEnd, GameEndSchema } from "./game-end/game-end";
import { InputCommand, InputCommandSchema } from "./input-command/input-command";
import { PilotSkillEffect, PilotSkillEffectSchema } from "./pilot-skill/pilot-skill-effect";
import { Reflect, ReflectSchema } from "./reflect/reflect";
import { RightItself, RightItselfSchema } from "./right-itself/right-itself";
import { StartGame, StartGameSchema } from "./start-game/start-game";
import { TurnChange, TurnChangeSchema } from "./turn-change/turn-change";
import { UpdateRemainingTurn, UpdateRemainingTurnSchema } from "./update-remaning-turn/update-remaining-turn";

/** ゲーム中に発生する効果をまとめたもの */
export type Effect =
  | StartGame
  | GameEnd
  | InputCommand
  | Reflect
  | BatteryDeclaration
  | Battle
  | TurnChange
  | BurstEffect
  | UpdateRemainingTurn
  | RightItself
  | PilotSkillEffect;

/** Effect zodスキーマ */
export const EffectSchema = z.union([
  StartGameSchema,
  GameEndSchema,
  InputCommandSchema,
  ReflectSchema,
  BatteryDeclarationSchema,
  BattleSchema,
  TurnChangeSchema,
  BurstEffectSchema,
  UpdateRemainingTurnSchema,
  RightItselfSchema,
  PilotSkillEffectSchema,
]);