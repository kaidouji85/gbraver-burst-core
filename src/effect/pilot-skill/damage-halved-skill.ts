import { DamageHalvedSkill } from "../../player/pilot/damage-halved-skill";
import { PlayerState } from "../../state/player-state";
import { PilotSkillInvokeParams } from "./pilot-skill-invoke-params";
import { PilotSkillInvokeResult } from "./pilot-skill-invoke-result";

/**
 * パイロットスキル発動者のステートを更新する
 * @param invoker パイロットスキル発動者のステート
 * @param skill パイロットスキル情報
 * @return パイロットスキル発動後のステート
 */
const updateInvoker = (
  invoker: PlayerState,
  skill: DamageHalvedSkill,
): PlayerState => ({
  ...invoker,
  armdozer: {
    ...invoker.armdozer,
    effects: [
      ...invoker.armdozer.effects,
      {
        type: "DamageHalved",
        period: {
          type: "TurnLimit",
          remainingTurn: skill.duration,
        },
      },
    ],
  },
});

/**
 * パイロットスキル ダメージ半減
 * @param params パイロットスキル発動情報
 * @return パイロットスキル発動結果
 */
export function damageHalvedSkill(
  params: PilotSkillInvokeParams<DamageHalvedSkill>,
): PilotSkillInvokeResult {
  const { invoker, other, skill } = params;
  return {
    invoker: updateInvoker(invoker, skill),
    other,
  };
}
