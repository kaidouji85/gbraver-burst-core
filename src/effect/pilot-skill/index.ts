import type { PlayerId } from "../../player/player";
import type { GameState, GameStateX } from "../../state/game-state";
import { PlayerState } from "../../state/player-state";
import { batteryBoost } from "./battery-boost";
import { batteryEnchantment } from "./battery-enchantment";
import { buffPower } from "./buff-power";
import { damageHalvedSkill } from "./damage-halved-skill";
import type { PilotSkillEffect } from "./pilot-skill-effect";
import { recoverBattery } from "./recover-battery";

/**
 * パイロットスキルを適用する
 * @param lastState 最新の状態
 * @param invokerId パイロットスキルを発動するプレイヤー
 * @return 更新結果、実行不可能な場合は例外を返す
 */
function pilotSkillEffect(
  lastState: GameState,
  invokerId: PlayerId,
): GameStateX<PilotSkillEffect> {
  const invoker = lastState.players.find((v) => v.playerId === invokerId);
  if (!invoker) {
    throw new Error("not found pilot skill invoker");
  }

  if (invoker.pilot.skill.type === "RecoverBatterySkill") {
    return recoverBattery(lastState, invokerId, invoker.pilot.skill);
  }

  if (invoker.pilot.skill.type === "BuffPowerSkill") {
    return buffPower(lastState, invokerId, invoker.pilot.skill);
  }

  if (invoker.pilot.skill.type === "BatteryEnchantmentSkill") {
    return batteryEnchantment(lastState, invokerId, invoker.pilot.skill);
  }

  if (invoker.pilot.skill.type === "DamageHalvedSkill") {
    return damageHalvedSkill(lastState, invokerId, invoker.pilot.skill);
  }

  if (invoker.pilot.skill.type === "BatteryBoostSkill") {
    return batteryBoost(lastState, invokerId, invoker.pilot.skill);
  }

  throw new Error("not found pilot skill");
}

/**
 * パイロットスキルを使用済みにする
 * @param lastState 最新状態
 * @return 更新結果
 */
function disablePilotSkill(
  lastState: GameStateX<PilotSkillEffect>,
): GameStateX<PilotSkillEffect> {
  const updateInvoker = (invoker: PlayerState) => ({
    ...invoker,
    pilot: { ...invoker.pilot, enableSkill: false },
  });
  const updatedPlayers = lastState.players.map((v) =>
    v.playerId === lastState.effect.invokerId ? updateInvoker(v) : v,
  );
  return { ...lastState, players: updatedPlayers };
}

/**
 * パイロットスキルを発動する
 * @param lastState 最新の状態
 * @param invokerId パイロットスキルを発動するプレイヤー
 * @return 更新結果、実行不可能な場合は例外を返す
 */
export function pilotSkill(
  lastState: GameState,
  invokerId: PlayerId,
): GameStateX<PilotSkillEffect> {
  const donePilotSkill = pilotSkillEffect(lastState, invokerId);
  return disablePilotSkill(donePilotSkill);
}
