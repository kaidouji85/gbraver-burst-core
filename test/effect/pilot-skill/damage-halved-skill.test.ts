import path from "path";

import {
  EMPTY_ARMDOZER_EFFECT,
  type GameState,
  type PlayerState,
} from "../../../src";
import { pilotSkill } from "../../../src/effect/pilot-skill";
import { EMPTY_GAME_STATE } from "../../../src/empty/game-state";
import { EMPTY_PILOT } from "../../../src/empty/pilot";
import { EMPTY_PLAYER_STATE } from "../../../src/empty/player";
import {
  exportSnapShotJSON,
  importSnapShotJSON,
  shouldUpdateSnapShot,
} from "../../snap-shot";
import {DamageHalvedSkill} from "../../../src/player/pilot/damage-halved-skill";

/** スキル ダメージ半減 */
const skill: DamageHalvedSkill = {
  type: "DamageHalvedSkill",
  duration: 1,
};

/** スキル発動者 */
const invoker: PlayerState = {
  ...EMPTY_PLAYER_STATE,
  playerId: "invoker",
  armdozer: {
    ...EMPTY_PLAYER_STATE.armdozer,
    effects: [EMPTY_ARMDOZER_EFFECT, EMPTY_ARMDOZER_EFFECT],
  },
  pilot: { ...EMPTY_PILOT, skill, enableSkill: true },
};

/** それ以外のプレイヤー */
const other: PlayerState = { ...EMPTY_PLAYER_STATE, playerId: "other" };

/** ゲームステート */
const state: GameState = {
  ...EMPTY_GAME_STATE,
  activePlayerId: invoker.playerId,
  players: [other, invoker],
};

test("ダメージ半減スキルが正しく発動できる", () => {
  const result = pilotSkill(state, invoker.playerId);
  const snapShotPath = path.join(__dirname, "damage-halved-skill.json");
  shouldUpdateSnapShot() && exportSnapShotJSON(snapShotPath, result);
  const snapShot = shouldUpdateSnapShot()
    ? result
    : importSnapShotJSON(snapShotPath);
  expect(result).toEqual(snapShot);
});
