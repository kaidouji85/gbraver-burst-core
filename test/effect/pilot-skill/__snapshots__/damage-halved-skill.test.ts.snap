// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ダメージ半減スキルが正しく発動できる 1`] = `
{
  "activePlayerId": "invoker",
  "effect": {
    "invokerId": "invoker",
    "name": "PilotSkillEffect",
    "skill": {
      "duration": 1,
      "type": "DamageHalvedSkill",
    },
  },
  "players": [
    {
      "armdozer": {
        "battery": 5,
        "burst": {
          "recoverBattery": 0,
          "turnStartBatteryCorrect": 0,
          "type": "RecoverBattery",
        },
        "effects": [],
        "enableBurst": true,
        "hp": 3000,
        "id": "id",
        "maxBattery": 5,
        "maxHp": 3000,
        "name": "name",
        "power": 2000,
        "speed": 2000,
      },
      "pilot": {
        "enableSkill": true,
        "id": "EMPTY_PLAYER",
        "name": "名無し",
        "skill": {
          "recoverBattery": 2,
          "type": "RecoverBatterySkill",
        },
      },
      "playerId": "other",
    },
    {
      "armdozer": {
        "battery": 5,
        "burst": {
          "recoverBattery": 0,
          "turnStartBatteryCorrect": 0,
          "type": "RecoverBattery",
        },
        "effects": [
          {
            "period": {
              "remainingTurn": 1,
              "type": "TurnLimit",
            },
            "type": "Empty",
          },
          {
            "period": {
              "remainingTurn": 1,
              "type": "TurnLimit",
            },
            "type": "Empty",
          },
          {
            "period": {
              "remainingTurn": 1,
              "type": "TurnLimit",
            },
            "type": "DamageHalved",
          },
        ],
        "enableBurst": true,
        "hp": 3000,
        "id": "id",
        "maxBattery": 5,
        "maxHp": 3000,
        "name": "name",
        "power": 2000,
        "speed": 2000,
      },
      "pilot": {
        "enableSkill": false,
        "id": "EMPTY_PLAYER",
        "name": "名無し",
        "skill": {
          "duration": 1,
          "type": "DamageHalvedSkill",
        },
      },
      "playerId": "invoker",
    },
  ],
}
`;
